### PR TITLE
Provide a Dockerfile to build nightly Docker images

### PR DESCRIPTION
**Pull request by michaelherger**
* to build: have a source folder with [`server`](http://github.com/Logitech/slimserver) and [`platform`](https://github.com/Logitech/slimserver-platforms) in it (yes, I know, stupid assumptions, but that's how the code was organized many years ago, when the `buildme.pl` script was first written).
* important TODO: `buildme.pl` lacks all of the uploading stuff. Maybe that's ok and can be done in the Jenkins job?
* added a `Slim/Utils/OS/Custom.pm` file to customize some of the LMS behaviour. Should be expanded to eg. enable the update checker etc.
* I'm sorry for the too many unrelated changes in `buildme.pl`. But oh well...

### CODE REVIEW

**Comment by michaelherger** [comment]
If you don't want to run the full `buildme.pl` script, just copy the `Docker` folder's content to the LMS root folder and build from there.

File: Docker/Custom-Slim-Utils-OS.pm [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+sub initPrefs {
+	my ($class, $prefs) = @_;
+
+	$prefs->{wizardDone} = 1;
+	$prefs->{libraryname} = Slim::Utils::Strings::string('SQUEEZEBOX_SERVER');
+	
+	if (-d MUSIC_DIR) {
+		$prefs->{mediadirs} = $prefs->{ignoreInImageScan} = $prefs->{ignoreInVidoeScan} = [ MUSIC_DIR ];
```
Comment (terual) [inline Docker/Custom-Slim-Utils-OS.pm]: Probably a typo in ignoreInVidoeScan

File: Docker/install.sh [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+# Configure user squeezeboxserver
+useradd squeezeboxserver
+usermod -u 99 squeezeboxserver
+usermod -g 100 squeezeboxserver
+usermod -d /home squeezeboxserver
+chown -R squeezeboxserver:users /home
+
+# Allow acces to /dev/snd
+usermod -a -G audio squeezeboxserver
```
Comment (terual) [inline Docker/install.sh]: Could we move this to the Dockerfile? There is not that much happening here anymore. Of course, we want to keep the number of layers low so it then should go into one RUN command.

**Comment by terual** [comment]
> If you don't want to run the full `buildme.pl` script, just copy the `Docker` folder's content to the LMS root folder and build from there.

The Dockerfile expects the source to reside in the directory server, but that's what buildme.pl prepares.
```Dockerfile
COPY server/ /opt/lms/
```
It would be nice if the Dockerfile can be used outside of this build environment, but we're not there yet I think. Would it also work if the docker files are copied in the buildme.pl script to $buildDir/server instead of $buildDir?

File: Docker/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	apt-get clean -qy && \
+	rm -rf /var/lib/apt/lists/* /tmp/* /var/tmp/*
+
+# Add & install LMS sript
+COPY install.sh /tmp/
+RUN chmod +x /tmp/install.sh
+RUN /tmp/install.sh
+
+COPY server/ /opt/lms/
+COPY Custom-Slim-Utils-OS.pm /opt/lms/Slim/Utils/OS/Custom.pm
```
Comment (terual) [inline Docker/Dockerfile]: It's a common practice in Docker to use a root directory for your app, it does not have to reside in /opt. We already do the same for /config, /music, etc.

File: Docker/install.sh (hunk 2 above)
Comment (devdems) [reply Docker/install.sh]: Agree, no need for this file anymore.

File: Docker/Custom-Slim-Utils-OS.pm (hunk 1 above)
Comment (michaelherger) [reply Docker/Custom-Slim-Utils-OS.pm]: Good catch!

File: Docker/install.sh (hunk 2 above)
Comment (michaelherger) [reply Docker/install.sh]: Thought so :-)

**Comment by terual** [comment]
@michaelherger, I believe you already created an org on Docker Hub. Is there already something there? As soon as something is there I would like to test with my LMS setup.

**Comment by michaelherger** [comment]
> @michaelherger, I believe you already created an org on Docker Hub. Is there already something there? As soon as something is there I would like to test with my LMS setup.

No, I failed to push there so far 🤦🏻. That's why there's still that big, fat "TODO" in `buildme.pl`.

**Comment by devdems** [comment]
> > @michaelherger, I believe you already created an org on Docker Hub. Is there already something there? As soon as something is there I would like to test with my LMS setup.
> 
> No, I failed to push there so far 🤦🏻. That's why there's still that big, fat "TODO" in `buildme.pl`.

Wouldn't be easier to do it in Git directly? Scroll to the bottom https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

**Comment by devdems** [comment]
> @michaelherger, I believe you already created an org on Docker Hub. Is there already something there? As soon as something is there I would like to test with my LMS setup.

After it's completed you should also contact hub team to get it verified. Afterward, the image will get this nice banner on it.
image

**Comment by michaelherger** [comment]
> > No, I failed to push there so far 🤦🏻. That's why there's still that big, fat "TODO" in `buildme.pl`.
> 
> Wouldn't be easier to do it in Git directly? Scroll to the bottom https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

You're probably right: `buildme.pl` isn't the right place. But I'll stick with the Jenkins host for now, as it's handling all the other builds, too.

Yet I have to figure out how to push correctly. It's a matter of time... I'll get there...

File: Docker/start-container.sh [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+usermod -o -u "$PUID" squeezeboxserver
+groupmod -o -g "$PGID" users
+
+umask 0002
+
+#Add permissions
+chown -R squeezeboxserver:users /config /playlist
+
+/usr/bin/perl /lms/slimserver.pl --user squeezeboxserver --prefsdir /config/prefs --logdir /config/logs --cachedir /config/cache
```
Comment (devdems) [inline Docker/start-container.sh:16]: What was the reason you removed supervisor? There is nothing controlling the service now.

Comment (michaelherger) [reply Docker/start-container.sh:16]: Is there a need for a supervisor? I never had any kind of supervisors in my LMS installations.

Comment (devdems) [reply Docker/start-container.sh:16]: For Docker installations it is. Most LMS dockers use images that already have a supervisor that is why you don't see them in the build. 

There should be a process that keeps an eye on the running services as if the break it will try to restart them.

It also feeds a user feedback in the docker log if everything is up and running.

Comment (terual) [reply Docker/start-container.sh:16]: I'm using an image for a while without a supervisor without problems. Looking up supervisord it is most of the time used for multiple services in one container, but that is here not really the case. Also, without supervisord the logs of LMS get directly feeded to the docker log. But if there is really a good reason for supervisor we should add it.

Comment (tmancill) [reply Docker/start-container.sh:16]: I have also used an image without a supervisor for a couple of years now without any problems.

In general, I think supervisors in single-process containers are a mistake.  They use additional memory, can confuse signal handling/propagation, and often hide the fact that a service is breaking - it's better to have the container exit, know that you have a problem, and fix the problem.  If you really need the process to be restarted no matter what, you can easily wrap your `docker run` in a systemd unit file (or some other supervisor running *outside* of docker).

Comment (devdems) [reply Docker/start-container.sh:16]: I'm looking at bigger picture to prepare a good base for future. For example for users that will need airplay we will have to run extra service.

Community would then need minor mod to "official" one as majority will be prepared.

Comment (michaelherger) [reply Docker/start-container.sh:16]: For me LMS has been so stable I never saw the need to have a supervisor. And on the one system where we actually have something (Debian), users are asking to get rid of it, for the reasons @tmancill mentioned: if things crash you should know about it. A supervisor which just tries to restart a failing service would often hide the problems. I'd rather know about them.

Plus I really try to keep this the simplest possible. Yes, there will be a need for more feature complete images, or even to extend this one. But before we go there I'd prefer to start simple, see how that's working out for people, then add if needed.

**Comment by michaelherger** [comment]
I got the multi-platform build working on my dev machine (Mac). But I can't get it to work on the build host (Debian). It's running Docker 19.03, I added the "experimental: true" flag to its configuration, and I have restarted the daemon. Yet calling buildx fails: "command buildx is not available" or similar. Any pointer?

Different topic: tags. Should tags be generic (`latest`), or is it ok to use revision numbers etc? Should we only have eg. `latest` (nightly 8.0), and probably a `8.0` once that's officially released? But then this wouldn't allow a user to roll back...

**Comment by terual** [comment]
Docker buildx seems to be only a part of the Docker Desktop distribution on Mac and Windows. I followed [these steps](https://github.com/docker/buildx#with-buildx-or-docker-1903) on my debian system with docker 19.03 to get the docker buildx command available.

**Comment by michaelherger** [comment]
> Docker buildx seems to be only a part of the Docker Desktop distribution on Mac and Windows. I followed [these steps](https://github.com/docker/buildx#with-buildx-or-docker-1903) on my debian system with docker 19.03 to get the docker buildx command available.

Oh, I read the documentation like "is included in 19.03". But it seems to require 19.03, and is only included in the desktop builds. Thanks! I'll give that a try.

What about the tagging?

**Comment by terual** [comment]
According to [docker best pratices](https://docs.docker.com/develop/dev-best-practices/) we should not only rely on the standard latest tag, but (also) use more descriptive tags. The tag 8.0-dev could be used until the release of 8.0. I don't think revision numbers for a development series makes sense. If you are tracking 8.0 then the most likely fix for your problem should be to update to latest instead of rolling back.

In short: I think you should use both latest as 8.0-dev for the current build. You can use multiple tags in the same build command by the way.

**Comment by terual** [comment]
> Oh, I read the documentation like "is included in 19.03".

Me too, it is very ambiguously written..

**Comment by mherger** [comment]
> According to [docker best pratices](https://docs.docker.com/develop/dev-best-practices/) we should not only rely on the standard latest tag, but (also) use more descriptive tags. The tag 8.0-dev could be used until the release of 8.0. I don't think revision numbers for a development series makes sense. If you are tracking 8.0 then the most likely fix for your problem should be to update to latest instead of rolling back.

But using a build revision number like we do for the other builds would require additional housekeeping to keep the number of tags under control. Anyone experienced with this?

> In short: I think you should use both latest as 8.0-dev for the current build. You can use multiple tags in the same build command by the way.

With `buildx` too? I failed. But then I've failed to do many things I tried related to Docker 😁

**Comment by terual** [comment]
> But using a build revision number like we do for the other builds would require additional housekeeping to keep the number of tags under control. Anyone experienced with this?

What I meant was **not** using revision numbers, but just one tag '8.0-dev' until the release. Or does using just one tag require extra work?

> With `buildx` too? I failed. But then I've failed to do many things I tried related to Docker grin

Doesn't adding multiple -t flags to the build command work? On my system it does. If you then subsequently push without specifying a tag, all tags are pushed.

**Comment by michaelherger** [comment]
> Doesn't adding multiple -t flags to the build command work? On my system it does. If you then subsequently push without specifying a tag, all tags are pushed.

Sure it does... I should probably call it a day.

**Comment by mherger** [comment]
It's time to merge this... thanks for all your input!

**Comment by terual** [comment]
I tried the latest docker image, but I got a permissions issue. The start-container script is run by root inside the container which changes the ownership of /config to squeezeboxserver:users, but then also runs the LMS service under the root user I think. This was one of the benefits of using supervisor, because you could specify the user under which the service runs.

**Comment by tmancill** [comment]
> I tried the latest docker image, but I got a permissions issue. The start-container script is run by root inside the container which changes the ownership of /config to squeezeboxserver:users, but then also runs the LMS service under the root user I think. This was one of the benefits of using supervisor, because you could specify the user under which the service runs.

I haven't switched over the official image, but the entry point should allow you to become whatever user you want - e.g., I use this locally:

```
exec sudo -u squeezeboxserver -E squeezeboxserver \
	--prefsdir $SQUEEZE_VOL/prefs \
	--logdir $SQUEEZE_VOL/logs \
	--cachedir $SQUEEZE_VOL/cache "$@"
```

**Comment by michaelherger** [comment]
> I tried the latest docker image, but I got a permissions issue. The start-container script is run by root inside the container which changes the ownership of /config to squeezeboxserver:users, but then also runs the LMS service under the root user I think. This was one of the benefits of using supervisor, because you could specify the user under which the service runs.

What permissions issue? How did you start it?

**Comment by terual** [comment]
What I did was first try the image:
```bash
docker pull lmscommunity/logitechmediaserver
docker run lmscommunity/logitechmediaserver
```
And got this output:
```
usermod: no changes
[20-10-23 05:59:57.3610] main::init (388) Starting Logitech Media Server (v8.0.0, 1603391272, Thu Oct 22 20:49:20 CEST 2020) perl 5.024001 - x86_64-linux-gnu-thread-multi
[20-10-23 05:59:57.4296] YAML::XS::DumpFile (31) Warning: Can't open '/config/cache/plugin-data.yaml' for output:
Permission denied at /lms/CPAN/YAML/XS.pm line 31.
[20-10-23 05:59:57.4301] Slim::Utils::Prefs::Namespace::savenow (337) Error: can't save /config/prefs/plugin/state.prefs: No such file or directory at /lms/Slim/Utils/Prefs/Namespace.pm line 326.
[20-10-23 05:59:57.4303] Slim::Utils::Prefs::Namespace::savenow (337) Error: can't save /config/prefs/server.prefs: Permission denied at /lms/Slim/Utils/Prefs/Namespace.pm line 326.
```

**Comment by michaelherger** [comment]
You didn't map the `/config` folder to the host? I indeed never tested this. And I am seeing some issues with a mapped `/music` on Synology, too. Did work on the dev machine.

**Comment by terual** [comment]
I think the issue is that LMS is run under the root user:
```
$ docker run -it --entrypoint id lmscommunity/logitechmediaserver
uid=0(root) gid=0(root) groups=0(root)
```

When I do this LMS starts:
```
$ docker run -it --entrypoint /bin/sh lmscommunity/logitechmediaserver 
# chown -R squeezeboxserver:users /config /playlist
# su - squeezeboxserver
$ /usr/bin/perl /lms/slimserver.pl --user squeezeboxserver --prefsdir /config/prefs --logdir /config/logs --cachedir /config/cache
```

**Comment by devdems** [comment]
Correct. @michaelherger you copied this run command from supervisor which has a different syntax.
